spec/elastic_apm/transaction_spec.rb

```diff
@@ -103,30 +103,28 @@ module ElasticAPM
     end
 
     describe '#inc_started_spans!' do
-      it 'increments count' do
-        expect { subject.inc_started_spans! }
-          .to change(subject, :started_spans).by 1
+      let!(:result) { subject.inc_started_spans! }
```

Comment:
Experience has taught me to stay away from `let!` with a bang. Personal preference only but I much prefer either using `before` or just inlining in each example with `result = ...`.

Reply:
I would advocate for using `let!` here because of these reasons
1) It's a construct of the `RSpec` framework, meant to be used exactly as it is here.
2) Setting local variables inside `it` block while there are test framework constructs to achieve the same functionality seems like an anti-pattern.
3) I'd like to have the two separate tests, one for the result of calling `subject.inc_started_spans!` and one for the counter change after `subject.inc_started_spans!`. If I use the `let!` block, the value of `result` is reused for each test.

Reply:
My experience is that excessive `let` use (as in much more than 1 or 2 per spec file) leads to less code for the cost of increased indirection. Especially when coming back to a test 6 months later, figuring out what goes on before getting to the actual `it` example can be a task on its own. This is why I'm often sacrificing the brevity over ease of understanding.

I'm not sure I would argue that _plain ruby_ eg. inline variables in this (or maybe any) case would be an anti-pattern.

These are just my thoughts and suggestions though. Keep it like this if this is your preference 😊

Reply:
I see your point about excessive let blocks getting confusing. I do think that if the `it` blocks are kept lean-- ideally to a single validation-- then the test can be very clear; all the setup is kept out of the validation and resources are not wasted as the `let`'s are memoized.
Anyway, let's just leave this as it is, as I think it makes a lot of sense. I'll be sure not to use `let`'s excessively though = )

Reply:
You're right. And I like rules like _only use `let!` if it makes sense to call it `:result`_ or something like that. That way it's indirection but indirection based on common knowledge in that project/context.

👍